商品出品機能

testarossa0214 opened the pull request:
# What
商品出品機能実装

# Why
ユーザーが商品を出品できるようにするため


【ログイン時状態の場合は、商品出品ページへ遷移できる動画】
https://gyazo.com/676be4523a0bd5f2502036b33bb9fe1b

【価格が入力されると同時に、販売手数料と販売利益が表示される動画】
https://gyazo.com/2a1ea5f447d97afdde2f7e5a02d41d93

【必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画】
https://gyazo.com/85254a6eedf30b055fef49a3161c0fea

【入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画】
https://gyazo.com/70054ef22318b07ca01c8eb664ccfd55

【ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画】
https://gyazo.com/ccb69dc2d376aef54f251c627722bb8c

【itemモデル単体テスト】
https://gyazo.com/d1b64508a734ab430786c37e1b378023

testarossa0214 committed:
itemモデル作成、マイグレーション

testarossa0214 committed:
アソシエーション記述

testarossa0214 committed:
active_storage導入

testarossa0214 committed:
出品ページへのルーティング、コントローラー、ビュー編集

testarossa0214 committed:
ActiveHash導入

testarossa0214 committed:
ActiveHash実装

testarossa0214 committed:
ActiveHash実装、JSファイル作成

testarossa0214 committed:
価格表示(JS記述)実装、画像バリデーション実装

testarossa0214 committed:
ログインページ変遷、販売金額バリデーション追加

testarossa0214 committed:
itemモデル単体テスト

testarossa0214 committed:
rubocop

testarossa0214 committed:
エラー文表示のためのビューファイル更新

Hide-Morishita commented on app/models/item.rb:
```diff
@@ -0,0 +1,25 @@
+    validates :product                                            # 商品名及び商品の説明欄はビューのmaxlengthで文字数を制限してるのでバリデーション記述していない
+    validates :product_description
+    validates :image
+    validates :price, format: { with: /\A[0-9]+\z/ },
+                      numericality: { only_integer: true,
+                                      greater_than: 300, less_than: 10_000_000 }
```
```ruby
greater_than: 300, less_than: 10_000_000
```
金額の範囲指定ですが、greater_thanの場合、**指定された値よりも大きくなければならないことを指定**します。
現状ですと300円ではバリデーションに弾かれて、保存できなくなってしまいますので修正してみましょう。

Hide-Morishita commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
-    <%# render 'shared/error_messages', model: f.object %>
+    
+    <%= render 'shared/error_messages', model: f.object %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```ruby
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
こちらは、実装案内用のコメントアウトですので削除しておきましょう。

Hide-Morishita commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,83 @@
+        expect(@item.errors.full_messages).to include("Price can't be blank", 'Price is invalid', 'Price is not a number')
+      end
+
+      it '価格が300未満だと出品できない' do
+        @item.price = '299'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be greater than 300')
+      end
+
+      it '価格が10000000以上だと出品できない' do
+        @item.price = '10000000'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be less than 10000000')
+      end
```
```ruby
@item.price = '299'
@item.price = '10000000'
```
priceはinteger型ですので、値は文字列ではなく数値にしましょう。
（シングルクォーテーションを外しましょう）

Hide-Morishita commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,83 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be less than 10000000')
+      end
+    end
```
販売価格に関してですが、より正確なテストを行うために、下記テストコードも実装しましょう。
```
・半角英数混合では出品できないこと
・半角英語だけでは出品できないこと
・全角文字では出品できないこと
```

Hide-Morishita replied:
異常系のテストに以下を追加しましょう。
```
ユーザー情報が紐付いていない場合は出品できないこと
```

Hide-Morishita reviewed (commented):
お待たせしました、レビュー依頼、キャプチャ動画ありがとうございます。
確認させていただきました。
コメントいたしましたので、ご確認をお願いいたします。
コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りいただくようお願いいたします。

testarossa0214 committed:
レビュー修正1

金額の範囲指定（300以下部分）修正
priceの値を数値へ（シングルクォーテーション外す）
テストコード追加
・半角英数混合のとき
・半角英語のとき
・全角のとき
・ユーザーが紐づいていないとき

testarossa0214 commented:
レビューありがとうございます。
上記内容を修正しましたので、ご確認のほどよろしくお願いいたします。

また、テストコード変更にともない新しい動画のリンクを添付いたします。
併せてご確認いただけますと幸いです。
https://gyazo.com/cbf80f5d6830b18ac28be5143d7bca01

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます。
出品機能の実装ですがLGTMとさせていただきます！
実装お疲れさまでした！！

マージ後、ローカルのmasterブランチでプルをお願いいたします。
その後masterブランチから新たなブランチを作成し、次の実装に取り組んでいただけますと幸いです。
次の実装も頑張っていきましょう！

testarossa0214 commented:
レビューありがとうございます！
次もがんばります！
